lib/brain.js

```diff
@@ -2601,84 +2601,160 @@ Brain.prototype._dispense = function _dispense () {
     })
 }
 
+Brain.prototype.waitForCollection = function waitForCollection (dispenseGenerator) {
+  const txId = this.tx.id
+  let g = dispenseGenerator.call(this) // change execution context with .call
+  let start = true
+  let yieldedVal
+
+  const checkCollection = () => {
+    const isCollecting = this.tx.id === txId && this.state === 'dispensing'
+
+    if (start) yieldedVal = g.next()
+    start = false
+
+    if (!isCollecting) {
+      return clearInterval(timerHandle)
+    }
+
+    this.billDispenser.billsPresent()
+      .then(billsPresent => {
+        if (!billsPresent) {
+          yieldedVal = g.next(yieldedVal)
+          emit('billDispenserCollected')
+        }
+      })
+  }
+
+  const timerHandle = setInterval(checkCollection, 500)
+  
+}
+
 Brain.prototype._physicalDispense = function _physicalDispense () {
+  const self = this
   const fiatCode = this.tx.fiatCode
   const txId = this.tx.id
+  let batchNum = 1
   const notes = [
     this.tx.bills[0].provisioned,
     this.tx.bills[1].provisioned
   ]
+  let notesToDispense = []
+
+  function optimize(inputArr, limit) {
+    if (limit === undefined) return [inputArr]
+
+    let inputHead = inputArr.slice(-2)[0]
+    let inputTail = inputArr.slice(-1)[0]
+    let optimizedArray = []
+  
+    function findLimitSum(a, b) {
+      let head = a
+      let tail = b
+      while ((head + tail) > limit) {
+        head = head === 0 ? 0 : --head
+        tail = tail === 0 ? 0 : --tail
+      }
+      optimizedArray.push([head, tail])
+      if ( a === 0 && b === 0) {
+          optimizedArray.pop()
+          return
+        }
+      findLimitSum(a - head, b - tail, limit)
+    }
+  
+    findLimitSum(inputHead, inputTail)
+
+    return optimizedArray
+  }
+
+  notesToDispense = optimize(notes, this.billDispenser.dispenseLimit)
 
   if (fiatCode !== this.billDispenser.fiatCode) {
     console.log('Wrong dispenser currency; dispenser: %s, tx: %s',
       this.billDispenser.fiatCode, fiatCode)
     return this._timedState('wrongDispenserCurrency')
   }
 
+  self.browser().send({dispenseBatch: {current: batchNum, of: notesToDispense.length} })
   this._transitionState('dispensing')
   emit('billDispenserDispensing')
+  
+  function* dispenseGenerator() {
+    let result = []
+    for (let index = 0; index < notesToDispense.length; index++) { // cannot yield inside forEach
+      let yielded = yield this.billDispenser.dispense(notesToDispense[index])
+      if (yielded !== undefined) result.push(yielded)
+      batchNum = index + 1
+      self.browser().send({dispenseBatch: {current: batchNum, of: notesToDispense.length} })
+    }
+    yield Promise.all(result.map(i => i.value))
+      .then(result => {
+        const bills = result.map(d => d.bills).reduce((p, c) => {
+          return [
```

Comment:
Need a unit test for this and for the ``optimize`` function.

Reply:
Got it